src/buffer/out/sgrStack.hpp

```diff
@@ -0,0 +1,238 @@
+            {
+                _numSgrPushes++;
+            }
+        }
+
+        // Method Description:
+        // - Restores text attributes by removing from the top of the internal stack,
+        //   combining them with the supplied currentAttributes, if appropriate.
+        // Arguments:
+        // - currentAttributes - The current text attributes. If only a portion of
+        //   attributes were saved on the internal stack, then those attributes will be
+        //   combined with the currentAttributes passed in to form the return value.
+        // Return Value:
+        // - The TextAttribute that has been removed from the top of the stack, possibly
+        //   combined with currentAttributes.
+        const TextAttribute Pop(const TextAttribute& currentAttributes) noexcept
+        {
+            if (_numSgrPushes > 0)
+            {
+                _numSgrPushes--;
+
+                if (_numSgrPushes < _countof(_storedSgrAttributes))
+                {
+// Must disable 26482 "Only index into arrays using constant expressions" because we are
+// implementing a stack, and that's the whole point.
+// We also disable the warning for using gsl::at, because doing that yields another: "No
+// array to pointer decay".
+#pragma warning(push)
+#pragma warning(disable : 26482 26446)
+                    const uint32_t validParts = _validAttributes[_numSgrPushes];
+
+                    if (validParts == UINT32_MAX)
+                    {
+                        return _storedSgrAttributes[_numSgrPushes];
+                    }
+                    else
+                    {
+                        return _CombineWithCurrentAttributes(currentAttributes,
+                                                             _storedSgrAttributes[_numSgrPushes],
+                                                             validParts);
+                    }
+#pragma warning(pop)
+                }
+            }
+
+            return currentAttributes;
+        }
+
+        // Xterm allows the save stack to go ten deep, so we'll follow suit. Pushes after
+        // ten deep will still remain "balanced"--once you pop back down below ten, you'll
+        // restore the appropriate text attributes. However, if you get more than a
+        // hundred pushes deep, we'll stop counting. Why unbalance somebody doing so many
+        // pushes? Putting a bound on it allows us to provide "reset" functionality: at
+        // any given point, you can execute 101 pops and know that you've taken the stack
+        // (push count) to zero. (Then you reset text attributes, and your state is
+        // clean.)
+        static constexpr int c_MaxStoredSgrPushes = 10;
+        static constexpr int c_MaxBalancedPushes = 100;
+
+    private:
+        static constexpr uint32_t _GraphicsOptionToFlag(DispatchTypes::GraphicsOptions option)
+        {
+            int iOption = static_cast<int>(option);
+
+            if (iOption < (sizeof(uint32_t) * 8))
```

Comment:
wait _what_ is this doing?

Reply:
I think this should be clearer now that I'm using a `std::bitset`.